### PR TITLE
Resurrect old color picker

### LINKED ISSUE #20415: Issues with the new colorpicker
Im opening this issue with the context of providing a centralized feedback for any issues that may arise with the new colorpicker.
All others please feel free to contribute your concerns and feedback with this upcoming change in the next release also.

Before, it was possible to conveniently input hex color code to provide teamcolor presets. With the change to hsv this has became no longer possible as it wants to arbitrarily restrict into a narrow range.
Further issues present in the process of converting a hex color into a format this new colorpicker understands, i would dare even say its not exactly human useable in its current state.

The intention was to make C4B060 pickable,
to that end have entered:
```
    ColorPickerManager:
        PreviewActor: fact
        PresetHues: 47
        PresetSaturations: 51
        HsvSaturationRange: 0.3, 0.97
        SimilarityThreshold: 0.16
```
this resulted in a entirely different color ingame:
coloringame
while this was expected:
colororig
The difference occured because the colorpicker is expecting a different hue, 0-255 rather than 0-360. 
Been told to multiply the obtained hue by 255 / 360, however that also resulted in a color other than the one wanted.
Later a close enough color #c2ac5f could be obtained by inputting 0.13 as it's hue.
tonediff
On the left the color obtained, on the right the color would like to get.

Another factor is since you can change only H and S, while V remains forcedly the same for all this contributed towards resulting in a different color alltogether.

I would deem this inacceptable in this time and year, you should able to get the color you want, that feels like wrestling with colormatic at paintshops, only to get a wholly different thing delivered anyhow.
Also should able to disable it alltogether like i can do before.

### PR DESCRIPTION
**Pull request by PunkPun**
Closes #20415

Credit to @darkademic. I just added a few of the missing bits such as the update rule.

Color picker in this PR
Screenshot 2023-03-12 at 20 37 22

Old color picker
Screenshot 2023-02-27 at 22 45 54

Bleed color picker
Screenshot 2023-02-27 at 22 45 40

As you can see the the color picker looks better with S and V axes flipped.

I also implemented `IColorPickerManagerInfo` interface so that third party mods would be able to implement their own color pickers. When reviewing the second commit I recommend to start with the interface.

### CODE REVIEW

File: OpenRA.Mods.Common/Traits/World/ColorPickerManager.cs [hunk 1]
```diff
@@ -61,15 +64,36 @@ public void RulesetLoaded(Ruleset rules, ActorInfo ai)
...
+		{
+			if (faction == null)
+			{
+				if (PreviewActor == null)
+					throw new YamlException($"{nameof(ColorPickerManager)} does not define a preview actor.");
+
+				return PreviewActor;
+			}
+
+			if (FactionPreviewActors.TryGetValue(faction, out var actorType))
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/World/ColorPickerManager.cs]: Add `faction != null` here then the check above can be removed.

**Comment by anvilvapre** [comment]
i wanted to mention that players in general don't mind if multiple players - in the same team or not - have about the same color - for as long as they are able to recognize and distinguise their own units. so if i.e. the progress bar for your own units would have a different color than of other players, or there is some mode/key that tells you to turn other players to another tint, that could work.

the issue is not that colors may look alike, but that you don't know which units are yours or not.

File: mods/cnc/rules/world.yaml [hunk 2]
```diff
@@ -271,6 +271,7 @@ World:
 		PreviewActor: fact.colorpicker
 		PresetHues: 0, 0.125, 0.185, 0.4, 0.54, 0.66, 0.79, 0.875, 0, 0.14, 0.23, 0.43, 0.54, 0.625, 0.77, 0.85
 		PresetSaturations: 0.9, 0.9, 0.9, 0.9, 0.9, 0.9, 0.9, 0.9, 0.4, 0.5, 0.4, 0.5, 0.4, 0.5, 0.4, 0.5
+		PresetValues: 0.95, 0.95, 0.95, 0.95, 0.95, 0.95, 0.95, 0.95, 0.95, 0.95, 0.95, 0.95, 0.95, 0.95, 0.95, 0.95
```
Comment (pchote) [inline mods/cnc/rules/world.yaml]: As mentioned in Discord, please change these to hex notation.

Can we please also replace the second row of desaturated colours in each mod with darkened colours instead?

Comment (PunkPun) [reply mods/cnc/rules/world.yaml]: The former would be for a follow up PR

**Comment by darkademic** [comment]
Just noticed the HueSliderWidget.cs has the old copyright header as it was based on prep-2211.

File: OpenRA.Mods.Common/Widgets/HueSliderWidget.cs [hunk 3]
```diff
@@ -0,0 +1,69 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2020 The OpenRA Developers (see AUTHORS)
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Widgets/HueSliderWidget.cs]: Wrong header.

File: OpenRA.Mods.Common/UpdateRules/Rules/20221203/AddColorPickerValueRange.cs [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+				yield break;
+
+			var v = manager.LastChildMatching("V")?.NodeValue<float>() ?? 0.95f;
+			manager.RemoveNodes("V");
+
+			var length = manager.LastChildMatching("PresetHues")?.NodeValue<float[]>().Length;
+			if (length == null)
+				yield break;
+
+			var vrray = new float[length.Value];
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/UpdateRules/Rules/20221203/AddColorPickerValueRange.cs]:
```suggestion
			var vArray = new float[length.Value];
```
(But this is in the update rule, so not important.)

File: OpenRA.Mods.Common/Traits/World/ColorPickerManager.cs [hunk 5]
```diff
@@ -61,15 +64,36 @@ public void RulesetLoaded(Ruleset rules, ActorInfo ai)
...
-		public IEnumerable<Color> PresetColors()
+		IEnumerable<Color> IColorPickerManagerInfo.PresetColors()
 		{
 			for (var i = 0; i < PresetHues.Length; i++)
-				yield return Color.FromAhsv(PresetHues[i], PresetSaturations[i], V);
+				yield return Color.FromAhsv(PresetHues[i], PresetSaturations[i], PresetValues[i]);
 		}
 
-		public Color Color;
+		Color IColorPickerManagerInfo.ColorPickerPaletteColor { get; set; }
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/World/ColorPickerManager.cs]: Why was this renamed? IMO `ColorPickerPaletteColor` is just adding unnecessary redundancy, (`colorManager.ColorPickerPaletteColor` vs `colorManager.Color`).

File: OpenRA.Mods.Common/TraitsInterfaces.cs [hunk 6]
```diff
@@ -290,6 +291,17 @@ public interface IProvidesAssetBrowserColorPickerPalettes
 		IEnumerable<string> ColorPickerPaletteNames { get; }
 	}
 
+	public interface IColorPickerManagerInfo : ITraitInfoInterface
+	{
+		Color ColorPickerPaletteColor { get; set; }
+		(float hMin, float hMax, float sMin, float sMax, float vMin, float vMax) GetHSVRange { get; }
```
Comment (pchote) [inline OpenRA.Mods.Common/TraitsInterfaces.cs]: IMO this is a really ugly method signature. Can we please split this into `GetHueRange`, `GetSaturationRange`, `GetValueRange`?

File: OpenRA.Mods.Common/TraitsInterfaces.cs [hunk 7]
```diff
@@ -290,6 +291,17 @@ public interface IProvidesAssetBrowserColorPickerPalettes
...
+		string GetPreviewActor(string faction = null);
```
Comment (pchote) [inline OpenRA.Mods.Common/TraitsInterfaces.cs]: Displaying an actor is a mod-specific implementation detail (a mod might want to recolor a static png image of a base, for example) so i'm not convinced that `GetPreviewActor` belongs on this interface.

File: OpenRA.Mods.Common/Widgets/ColorMixerWidget.cs [hunk 8]
```diff
@@ -66,19 +68,23 @@ public void SetColorLimits(float minSaturation, float maxSaturation, float v)
 				fixed (byte* cc = &buffer[0])
 				{
 					var c = (int*)cc;
-					for (var s = 0; s < 256; s++)
-						for (var h = 0; h < 256; h++)
+					for (var v = 0; v < 256; v++)
```
Comment (pchote) [inline OpenRA.Mods.Common/Widgets/ColorMixerWidget.cs:74]: Can we please add a set of braces here so we don't have a hanging block?

File: OpenRA.Mods.Common/TraitsInterfaces.cs (hunk 6 above)
Comment (abcdefg30) [reply OpenRA.Mods.Common/TraitsInterfaces.cs]: We could add a record or struct for this instead.

File: OpenRA.Mods.Common/Traits/World/ColorPickerManager.cs (hunk 5 above)
Comment (PunkPun) [reply OpenRA.Mods.Common/Traits/World/ColorPickerManager.cs]: This color is just used for the color picker palette. An indirect way of passing a value. I don't like it, but IMO at the minimum the name should reflect it

File: OpenRA.Mods.Common/TraitsInterfaces.cs (hunk 7 above)
Comment (PunkPun) [reply OpenRA.Mods.Common/TraitsInterfaces.cs]: I'm not sure how to abstract it then. The definition needs to be transferred to another trait / widget.

File: OpenRA.Mods.Common/Traits/World/ColorPickerManager.cs (hunk 1 above)
Comment (PunkPun) [reply OpenRA.Mods.Common/Traits/World/ColorPickerManager.cs]: it can't, this is a very specific implementation. Tough I could make it less robust to make it look cleaner

File: OpenRA.Mods.Common/TraitsInterfaces.cs (hunk 7 above)
Comment (PunkPun) [reply OpenRA.Mods.Common/TraitsInterfaces.cs]: perhaps add a `ActorColorPickerPreviewWidget : ActorPreviewWidget`

**Comment by PunkPun** [comment]
Made the IColorPickerManagerInfo interface much more generic and cleaned up a bunch of code in the process

File: OpenRA.Mods.Common/Traits/Palettes/ColorPickerPalette.cs [hunk 9]
```diff
@@ -44,19 +44,22 @@ class ColorPickerPaletteInfo : TraitInfo
 	class ColorPickerPalette : ILoadsPalettes, IProvidesAssetBrowserColorPickerPalettes, ITickRender
 	{
 		readonly ColorPickerPaletteInfo info;
-		readonly ColorPickerManagerInfo colorManager;
 		Color color;
+		Color preferedColor;
```
Comment (Mailaender) [inline OpenRA.Mods.Common/Traits/Palettes/ColorPickerPalette.cs]:
```suggestion
		Color preferredColor;
```

**Comment by Mailaender** [comment]
Changing to a predefined color and then switching back to mixer does not update properly:

image

**Comment by PunkPun** [comment]
fixed

**Review by Mailaender — DISMISSED** [review]

**Comment by Mailaender** [comment]
There is a superflous dot in the first commit message.

**Comment by PunkPun** [comment]
If you so wish @abcdefg30

File: OpenRA.Mods.Common/Traits/World/ColorPickerManager.cs [hunk 10]
```diff
@@ -34,17 +37,14 @@ public class ColorPickerManagerInfo : TraitInfo<ColorPickerManager>, IRulesetLoa
...
 		[Desc("Perceptual color threshold for determining whether two colors are too similar.")]
 		public readonly float SimilarityThreshold = 0.314f;
 
-		[Desc("List of hue components for the preset colors in the palette tab. Each entry must have a corresponding PresetSaturations definition.")]
-		public readonly float[] PresetHues = Array.Empty<float>();
-
-		[Desc("List of saturation components for the preset colors in the palette tab. Each entry must have a corresponding PresetHues definition.")]
-		public readonly float[] PresetSaturations = Array.Empty<float>();
+		[Desc("List of colors to be displayed in the palette tab.")]
+		public readonly Color[] PresetColors = Array.Empty<Color>();
```
Comment (penev92) [inline OpenRA.Mods.Common/Traits/World/ColorPickerManager.cs:47]: Should we be checking when loading these colors if they violate the Saturation or ValueRanges above?

File: OpenRA.Mods.Common/Traits/World/ColorPickerManager.cs [hunk 11]
```diff
@@ -179,8 +127,100 @@ Color MakeValid(float hue, float sat, MersenneTwister random, List<(float R, flo
 
 			// Failed to find a solution within a reasonable time: return a random color without any validation
 			onError?.Invoke(InvalidPlayerColor);
-			return Color.FromAhsv(random.NextFloat(), float2.Lerp(HsvSaturationRange[0], HsvSaturationRange[1], random.NextFloat()), V);
+			var randomSat = float2.Lerp(HsvSaturationRange[0], HsvSaturationRange[1], random.NextFloat());
+			var randomVal = float2.Lerp(HsvSaturationRange[0], HsvSaturationRange[1], random.NextFloat());
```
Comment (penev92) [inline OpenRA.Mods.Common/Traits/World/ColorPickerManager.cs]:
```suggestion
			var randomVal = float2.Lerp(HsvValueRange[0], HsvValueRange[1], random.NextFloat());
```

File: OpenRA.Mods.Common/UpdateRules/Rules/20221203/AddColorPickerValueRange.cs [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+#endregion
+
+using System.Collections.Generic;
+using OpenRA.Primitives;
+
+namespace OpenRA.Mods.Common.UpdateRules.Rules
+{
+	public class AddColorPickerValueRange : UpdateRule
+	{
+		public override string Name => "PresetHues, PresetSaturations and V was removed from ColorPickerManager, PresetColors added in their place.";
```
Comment (penev92) [inline OpenRA.Mods.Common/UpdateRules/Rules/20221203/AddColorPickerValueRange.cs]:
```suggestion
		public override string Name => "ColorPickerManager's PresetHues, PresetSaturations and V were replaced with PresetColors.";
```

File: OpenRA.Mods.Common/UpdateRules/Rules/20221203/AddColorPickerValueRange.cs [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+
+		public override string Description =>
+			"Each preset color can now have their brightness specified";
```
Comment (penev92) [inline OpenRA.Mods.Common/UpdateRules/Rules/20221203/AddColorPickerValueRange.cs]:
```suggestion
			"Each preset color can now have their brightness specified.";
```

File: OpenRA.Mods.Common/UpdateRules/Rules/20221203/AddColorPickerValueRange.cs [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+			if (manager == null)
+				yield break;
+
+			var v = manager.LastChildMatching("V")?.NodeValue<float>() ?? 0.95f;
+			manager.RemoveNodes("V");
+
+			var hues = manager.LastChildMatching("PresetHues")?.NodeValue<float[]>();
+			var saturations = manager.LastChildMatching("PresetSaturations")?.NodeValue<float[]>();
+			if (hues == null || saturations == null)
+				yield break;
```
Comment (penev92) [inline OpenRA.Mods.Common/UpdateRules/Rules/20221203/AddColorPickerValueRange.cs]: You already removed `V` before this point, which will leave broken (half-updated) YAML.
Please move `manager.RemoveNodes("V");` below this line where you remove the other nodes.

File: OpenRA.Mods.Common/Traits/World/ColorPickerManager.cs [hunk 15]
```diff
@@ -179,8 +127,100 @@ Color MakeValid(float hue, float sat, MersenneTwister random, List<(float R, flo
...
+			return Color.FromAhsv(random.NextFloat(), randomSat, randomVal);
+		}
+
+		#region IColorPickerManagerInfo
+
+		public event Action<Color> OnColorPickerColorUpdate;
+
+		(float hMin, float hMax) IColorPickerManagerInfo.GetHueRange { get; }
```
Comment (penev92) [inline OpenRA.Mods.Common/Traits/World/ColorPickerManager.cs]: This is never assigned to?

File: OpenRA.Mods.Common/TraitsInterfaces.cs [hunk 16]
```diff
@@ -290,6 +292,19 @@ public interface IProvidesAssetBrowserColorPickerPalettes
 		IEnumerable<string> ColorPickerPaletteNames { get; }
 	}
 
+	public interface IColorPickerManagerInfo : ITraitInfoInterface
+	{
+		(float hMin, float hMax) GetHueRange { get; }
+		(float sMin, float sMax) GetSaturationRange { get; }
+		(float vMin, float vMax) GetValueRange { get; }
```
Comment (penev92) [inline OpenRA.Mods.Common/TraitsInterfaces.cs]: FWIW "Get*" suggests a method, not a property.

File: OpenRA.Mods.Common/Traits/World/ColorPickerManager.cs [hunk 17]
```diff
@@ -179,8 +127,100 @@ Color MakeValid(float hue, float sat, MersenneTwister random, List<(float R, flo
...
+			return MakeValid(h, s, v, random, terrainColors, playerColors, null);
+		}
+
+		void IColorPickerManagerInfo.ShowColorDropDown(DropDownButtonWidget dropdownButton, Color initialColor, string initialFaction, WorldRenderer worldRenderer, Action<Color> onExit)
+		{
+			dropdownButton.RemovePanel();
+
+			void AddActorPreview(Widget parent)
+			{
+				var preview = parent.GetOrNull<ActorPreviewWidget>("PREVIEW");
```
Comment (penev92) [inline OpenRA.Mods.Common/Traits/World/ColorPickerManager.cs:186]: Do we have a precedent anywhere of traits managing widgets like that?

File: OpenRA.Mods.Common/Traits/World/ColorPickerManager.cs [hunk 18]
```diff
@@ -179,8 +127,100 @@ Color MakeValid(float hue, float sat, MersenneTwister random, List<(float R, flo
...
+				if (preview == null)
+					return;
+
+				if (initialFaction == null || !FactionPreviewActors.TryGetValue(initialFaction, out var actorType))
+				{
+					if (PreviewActor == null)
+						throw new YamlException($"{nameof(ColorPickerManager)} does not define a preview actor" + initialFaction == null ? "." : $"for faction {initialFaction}.");
```
Comment (penev92) [inline OpenRA.Mods.Common/Traits/World/ColorPickerManager.cs]: This needs some brackets to evaluate properly:
```suggestion
						throw new YamlException($"{nameof(ColorPickerManager)} does not define a preview actor" + (initialFaction == null ? "." : $"for faction {initialFaction}."));
```

**Review by penev92 — CHANGES_REQUESTED** [review]
We still seem to be missing some colors very close to white, but I don't think that was ever the goal. We do now have so *very* dark colors.
Bonus points for yaml hex colors actually matching their in-game representations (making that particular ORAIDE feature useful again). 👍 

Couldn't spot issues ingame. I went over the first commit and got halfway through the second, so this isn't a complete review, but I still have a few comments to work on.

File: OpenRA.Mods.Common/Traits/World/ColorPickerManager.cs (hunk 10 above)
Comment (PunkPun) [reply OpenRA.Mods.Common/Traits/World/ColorPickerManager.cs:47]: They are going to get fixed anyway. But we could add a lint check if you feel it's needed.

File: OpenRA.Mods.Common/UpdateRules/Rules/20221203/AddColorPickerValueRange.cs (hunk 14 above)
Comment (PunkPun) [reply OpenRA.Mods.Common/UpdateRules/Rules/20221203/AddColorPickerValueRange.cs]: This behaviour is intentional as the others are not guaranteed to exist

File: OpenRA.Mods.Common/Traits/World/ColorPickerManager.cs (hunk 15 above)
Comment (PunkPun) [reply OpenRA.Mods.Common/Traits/World/ColorPickerManager.cs]: Yes, this is just the implementation of the interface. In our setup we use hue slider and don't limit hue

File: OpenRA.Mods.Common/UpdateRules/Rules/20221203/AddColorPickerValueRange.cs (hunk 14 above)
Comment (PunkPun) [reply OpenRA.Mods.Common/UpdateRules/Rules/20221203/AddColorPickerValueRange.cs]: I moved them up higher

File: OpenRA.Mods.Common/TraitsInterfaces.cs (hunk 16 above)
Comment (PunkPun) [reply OpenRA.Mods.Common/TraitsInterfaces.cs]: redacted

File: OpenRA.Mods.Common/Traits/World/ColorPickerManager.cs (hunk 15 above)
Comment (PunkPun) [reply OpenRA.Mods.Common/Traits/World/ColorPickerManager.cs]: tho I suppose it doesn't make any sense to have this. If other implementations need it they will extend the interface

removed

File: OpenRA.Mods.Common/Traits/World/ColorPickerManager.cs (hunk 17 above)
Comment (PunkPun) [reply OpenRA.Mods.Common/Traits/World/ColorPickerManager.cs:186]: Not that I'm aware of. Although not the most beautiful, this was the only way I found to not hardcode the preview

File: OpenRA.Mods.Common/Traits/World/ColorPickerManager.cs (hunk 10 above)
Comment (penev92) [reply OpenRA.Mods.Common/Traits/World/ColorPickerManager.cs:47]: Lint checks are always nice, but I wouldn't say it's *required* here. I was talking about runtime and as long as these colors are fixed to valid ranges that should be fine.

File: OpenRA.Mods.Common/Traits/World/ColorPickerManager.cs (hunk 15 above)
Comment (penev92) [reply OpenRA.Mods.Common/Traits/World/ColorPickerManager.cs]: Well actually not needing it is an implementation detail specific to our implementation. The interface makes sense to have it, being a supposedly generic one. If we don't need it, then I think something like this would be best
```cs
// Our implementation doesn't need or use this because we get the hue from the HueSliderWidget.
(float hMin, float hMax) IColorPickerManagerInfo.GetHueRange => (0, 0);
```

Comment (PunkPun) [reply OpenRA.Mods.Common/Traits/World/ColorPickerManager.cs]: I've already made up my mind. It makes 0 sense to have it.

* We don't use it
* The color manager doesn't use it
* Hue range restrictions look completely different. Restricting a circle between 2 values does not make any sense

Comment (pchote) [reply OpenRA.Mods.Common/Traits/World/ColorPickerManager.cs]: It doesn't make sense to have a hue range, as it does not have a natural min or max value (its circular).

File: OpenRA.Mods.Common/Widgets/Logic/ColorPickerLogic.cs [hunk 19]
```diff
@@ -29,14 +29,24 @@ public class ColorPickerLogic : ChromeLogic
...
 
 			// Set the initial state
 			// All users need to use the same TraitInfo instance, chosen as the default mod rules
 			var colorManager = modData.DefaultRules.Actors[SystemActors.World].TraitInfo<ColorPickerManagerInfo>();
-			mixer.SetColorLimits(colorManager.HsvSaturationRange[0], colorManager.HsvSaturationRange[1], colorManager.V);
+			mixer.SetColorLimits(colorManager.HsvSaturationRange[0], colorManager.HsvSaturationRange[1], colorManager.HsvValueRange[0], colorManager.HsvValueRange[1]);
 			mixer.OnChange += () => onChange(mixer.Color);
 			mixer.Set(initialColor);
 
+			hueSlider.OnChange += (h) =>
```
Comment (pchote) [inline OpenRA.Mods.Common/Widgets/Logic/ColorPickerLogic.cs]: Remove redundant `()`.

File: OpenRA.Mods.Common/Widgets/Logic/ColorPickerLogic.cs [hunk 20]
```diff
@@ -29,14 +29,24 @@ public class ColorPickerLogic : ChromeLogic
...
+			{
+				mixer.SetColorLimits(colorManager.HsvSaturationRange[0], colorManager.HsvSaturationRange[1], colorManager.HsvValueRange[0], colorManager.HsvValueRange[1], h);
+				mixer.Set(Color.FromAhsv(h, mixer.Color.ToAhsv().S, mixer.Color.ToAhsv().V));
```
Comment (pchote) [inline OpenRA.Mods.Common/Widgets/Logic/ColorPickerLogic.cs]:
```csharp
var (_, _, s, v) = mixer.Color.ToAhsv();
mixer.Set(Color.FromAhsv(h, s, v));
```

File: OpenRA.Mods.Common/Traits/Palettes/ColorPickerPalette.cs [hunk 21]
```diff
@@ -44,19 +44,22 @@ class ColorPickerPaletteInfo : TraitInfo
...
 
 		public ColorPickerPalette(ColorPickerPaletteInfo info)
 		{
-			// All users need to use the same TraitInfo instance, chosen as the default mod rules
-			colorManager = Game.ModData.DefaultRules.Actors[SystemActors.World].TraitInfo<ColorPickerManagerInfo>();
 			this.info = info;
+
+			// All users need to use the same TraitInfo instance, chosen as the default mod rules
+			var colorManager = Game.ModData.DefaultRules.Actors[SystemActors.World].TraitInfo<IColorPickerManagerInfo>();
+			colorManager.OnColorPickerColorUpdate += (c) => preferredColor = c;
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Palettes/ColorPickerPalette.cs]: Remove redundant `()`

File: OpenRA.Mods.Common/Traits/World/ColorPickerManager.cs (hunk 17 above)
Comment (pchote) [reply OpenRA.Mods.Common/Traits/World/ColorPickerManager.cs:186]: :thisIsFine:

This is a speciality heavily-ui-related trait, so I don't think this is a major issue.

It would be helpful, though, to have a comment explaining that this is done so that the actor preview stuff can remain encapsulated within this implementation.

File: OpenRA.Mods.Common/Widgets/Logic/AssetBrowserLogic.cs [hunk 22]
```diff
@@ -177,16 +177,17 @@ public AssetBrowserLogic(Widget widget, Action onExit, ModData modData, WorldRen
...
-			colorManager.Color = Game.Settings.Player.Color;
+			var colorManager = modData.DefaultRules.Actors[SystemActors.World].TraitInfo<IColorPickerManagerInfo>();
 
 			var colorDropdown = panel.GetOrNull<DropDownButtonWidget>("COLOR");
 			if (colorDropdown != null)
 			{
+				var color = Game.Settings.Player.Color;
 				colorDropdown.IsDisabled = () => !colorPickerPalettes.Contains(currentPalette);
-				colorDropdown.OnMouseDown = _ => ColorPickerLogic.ShowColorDropDown(colorDropdown, colorManager, worldRenderer);
+				colorDropdown.OnMouseDown = _ => colorManager.ShowColorDropDown(colorDropdown, color, null, worldRenderer, (c) => color = c);
```
Comment (pchote) [inline OpenRA.Mods.Common/Widgets/Logic/AssetBrowserLogic.cs]: Remove `()`.

File: OpenRA.Mods.Common/Widgets/Logic/ColorPickerLogic.cs [hunk 23]
```diff
@@ -25,22 +24,25 @@ public class ColorPickerLogic : ChromeLogic
...
 			var mixer = widget.Get<ColorMixerWidget>("MIXER");
 			var hueSlider = widget.Get<HueSliderWidget>("HUE_SLIDER");
 
 			// Set the initial state
 			// All users need to use the same TraitInfo instance, chosen as the default mod rules
-			var colorManager = modData.DefaultRules.Actors[SystemActors.World].TraitInfo<ColorPickerManagerInfo>();
-			mixer.SetColorLimits(colorManager.HsvSaturationRange[0], colorManager.HsvSaturationRange[1], colorManager.HsvValueRange[0], colorManager.HsvValueRange[1]);
+			var colorManager = modData.DefaultRules.Actors[SystemActors.World].TraitInfo<IColorPickerManagerInfo>();
+
+			(var sMin, var sMax) = colorManager.SaturationRange;
```
Comment (pchote) [inline OpenRA.Mods.Common/Widgets/Logic/ColorPickerLogic.cs]: `var (sMin, sMax)`

File: OpenRA.Mods.Common/Widgets/Logic/IntroductionPromptLogic.cs [hunk 24]
```diff
@@ -114,11 +111,13 @@ public IntroductionPromptLogic(Widget widget, ModData modData, WorldRenderer wor
 
 			SettingsUtils.BindCheckboxPref(widget, "EDGESCROLL_CHECKBOX", gs, "ViewportEdgeScroll");
 
+			var colorManager = modData.DefaultRules.Actors[SystemActors.World].TraitInfo<IColorPickerManagerInfo>();
+
 			var colorDropdown = widget.Get<DropDownButtonWidget>("PLAYERCOLOR");
 			colorDropdown.IsDisabled = () => worldRenderer.World.Type != WorldType.Shellmap;
-			colorDropdown.OnMouseDown = _ => ColorPickerLogic.ShowColorDropDown(colorDropdown, colorManager, worldRenderer, () =>
+			colorDropdown.OnMouseDown = _ => colorManager.ShowColorDropDown(colorDropdown, ps.Color, null, worldRenderer, (color) =>
```
Comment (pchote) [inline OpenRA.Mods.Common/Widgets/Logic/IntroductionPromptLogic.cs]: Remove `()`

File: OpenRA.Mods.Common/Widgets/Logic/Lobby/LobbyUtils.cs [hunk 25]
```diff
@@ -544,13 +519,22 @@ void OkPressed()
...
 
-		public static void SetupEditableColorWidget(Widget parent, Session.Slot s, Session.Client c, OrderManager orderManager, WorldRenderer worldRenderer, ColorPickerManagerInfo colorManager)
+		public static void SetupEditableColorWidget(Widget parent, Session.Slot s, Session.Client c, OrderManager orderManager, WorldRenderer worldRenderer, IColorPickerManagerInfo colorManager)
 		{
-			var color = parent.Get<DropDownButtonWidget>("COLOR");
-			color.IsDisabled = () => (s != null && s.LockColor) || orderManager.LocalClient.IsReady;
-			color.OnMouseDown = _ => ShowColorDropDown(color, c, orderManager, worldRenderer, colorManager);
+			var colorDropdown = parent.Get<DropDownButtonWidget>("COLOR");
+			colorDropdown.IsDisabled = () => (s != null && s.LockColor) || orderManager.LocalClient.IsReady;
+			colorDropdown.OnMouseDown = _ => colorManager.ShowColorDropDown(colorDropdown, c.Color, c.Faction, worldRenderer, (color) =>
```
Comment (pchote) [inline OpenRA.Mods.Common/Widgets/Logic/Lobby/LobbyUtils.cs]: Remove `()`

File: OpenRA.Mods.Common/Widgets/Logic/Settings/DisplaySettingsLogic.cs [hunk 26]
```diff
@@ -268,14 +268,13 @@ Func<bool> InitPanel(Widget panel)
...
 			};
 
-			var colorManager = modData.DefaultRules.Actors[SystemActors.World].TraitInfo<ColorPickerManagerInfo>();
-			colorManager.Color = ps.Color;
+			var colorManager = modData.DefaultRules.Actors[SystemActors.World].TraitInfo<IColorPickerManagerInfo>();
 
 			var colorDropdown = panel.Get<DropDownButtonWidget>("PLAYERCOLOR");
 			colorDropdown.IsDisabled = () => worldRenderer.World.Type != WorldType.Shellmap;
-			colorDropdown.OnMouseDown = _ => ColorPickerLogic.ShowColorDropDown(colorDropdown, colorManager, worldRenderer, () =>
+			colorDropdown.OnMouseDown = _ => colorManager.ShowColorDropDown(colorDropdown, ps.Color, null, worldRenderer, (color) =>
```
Comment (pchote) [inline OpenRA.Mods.Common/Widgets/Logic/Settings/DisplaySettingsLogic.cs]: Remove `()`

**Comment by penev92** [comment]
Sounds like this should finally close #20415.

File: OpenRA.Mods.Common/Traits/World/ColorPickerManager.cs (hunk 17 above)
Comment (PunkPun) [reply OpenRA.Mods.Common/Traits/World/ColorPickerManager.cs:186]: added a few comments

**Review by penev92 — APPROVED** [review]

**Comment by penev92** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/1c9f11785295a1e0566e5ab4d8b88e581a312e15...2376bee0c1411266ba31713be4158c02f95c2068)

Note: PR NOT cherry-picked to `prep-2211` pending decision on how to proceed.

**Comment by Mailaender** [comment]
I vote for picking to `prep-2211`.
